### PR TITLE
Laboratory work #3, Anna Podpryatova - 22FPL2

### PR DESCRIPTION
**Pull request by AnPruch**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+
+        tokens = []
+        for token in text.lower():
+            if token == ' ' and tokens[-1] != self._end_of_word_token:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:51]: `token == ' '` better change to `token` - it is a more pythonic way

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:51]: I don't understand you... How is this an equivalent of space check? Can you elaborate on it?

Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:51]: Ok, got it

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -456,7 +700,7 @@ class BackOffGenerator:
 
     Attributes:
         _language_models (dict[int, NGramLanguageModel]): Language models for next token prediction
-        _text_processor (NGramLanguageModel): A TextProcessor instance to handle text processing
+        _text_processor (TextProcessor): A TextProcessor instance to handle text processing
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:703]: Why did you change it?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good. Let's try 10!

File: lab_3_generate_by_ngrams/README.rst [hunk 3]
```diff
@@ -1299,10 +1299,6 @@ BackOff, мы предлагаем Вам расширить хранилище
 и начало последовательности в строковом виде, возвращает итоговую последовательность
 в виде строки.
 
-.. note:: Если на вход подаются некорректные значения (количество букв для
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/README.rst:1302]: Is this ok that the exact same note is repeated at 1326?

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -414,6 +620,7 @@ class NGramLanguageModelReader:
         _json_path (str): Local path to assets file
         _eow_token (str): Special token for text processor
         _text_processor (TextProcessor): A TextProcessor instance to handle text processing
+        _content (dict): ngrams from external JSON
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:623]: No, but in Readme file it was specified to be an attribute and I thought it just was forgotten in this listing. I was confused a lot because of it

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:703]: It seems wrong to have a text processor of a NGramLanguageModel type, it just interfers with the description, because it is by default TextProcessor type. Or isn't it? I didn't know if it's a mistake or I just don't understand the types of these inputs, so I wanted to draw attention and ask.

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -104,6 +146,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:152]: Thank you

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -354,9 +521,9 @@ class BeamSearchTextGenerator:
 
     Attributes:
         _language_model (tuple[NGramLanguageModel]): Language models for next token prediction
-        _text_processor (NGramLanguageModel): A TextProcessor instance to handle text processing
-        _beam_width (NGramLanguageModel): Beam width parameter for generation
-        beam_searcher (NGramLanguageModel): Searcher instances for each language model
+        _text_processor (TextProcessor): A TextProcessor instance to handle text processing
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:524]: Are the attributes really of this type? Readme file says specifically that they are TextProcessor and BeamSearcher types, so I was very confused

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -159,6 +233,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not (isinstance(decoded_corpus, tuple) and decoded_corpus):
+            return None
+        decoded_text = decoded_corpus[0].upper() + ''.join(decoded_corpus)[1:-1]
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Rewrote

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -159,6 +233,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        decoded_text = decoded_text.replace(self._end_of_word_token, ' ')
+        if decoded_corpus[-1] != self._end_of_word_token:
+            decoded_text += decoded_corpus[-1]
+        return decoded_text + '.'
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Rewrote

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -195,6 +280,10 @@ def set_n_grams(self, frequencies: dict) -> None:
         Args:
             frequencies (dict): Computed in advance frequencies for n-grams
         """
+        if not (isinstance(frequencies, dict) and frequencies):
+            return None
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Done

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -195,6 +280,10 @@ def set_n_grams(self, frequencies: dict) -> None:
...
+        self._n_gram_frequencies = frequencies
+        return None
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Done

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -76,6 +99,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        for key, identificator in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:104]: use `filter`. Better with `next` to get only first element

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -41,6 +45,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+
+        tokens = []
+        for token in text.lower():
+            if token.isspace() and tokens and tokens[-1] != self._end_of_word_token:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: too dificult to understand the condition

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -129,6 +183,11 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if isinstance(content, dict) and content:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:186]: :O i in for in for in if.... Please, simplify

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -159,6 +229,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not (isinstance(decoded_corpus, tuple) and decoded_corpus):
+            return None
+        if decoded_corpus[-1] != self._end_of_word_token:
+            return f"{''.join(decoded_corpus).capitalize()}."\
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:235]: looks very simular returns.. try to reuse

File: lab_3_generate_by_ngrams/main.py [hunk 15]
```diff
@@ -221,6 +328,13 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
 
         In case of corrupt input arguments, None is returned
         """
+        if not (isinstance(sequence, tuple) and sequence
+                and len(sequence) >= self._n_gram_size - 1):
+            return None
+        context = sequence[-self._n_gram_size + 1:]
+        return {n_gram[-1]: freq
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:335]: use `dict(filter(...))`

File: lab_3_generate_by_ngrams/main.py [hunk 16]
```diff
@@ -270,6 +390,26 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+        encoded = self._text_processor.encode(prompt)
+        if not (encoded and n_gram_size):
+            return None
+
+        for iteration in range(seq_len):
+            tokens = self._model.generate_next_token(encoded[-n_gram_size + 1:])
+            if not tokens:
+                break
+            max_freq = max(tokens.values())
+            candidates_max = [candidate for candidate, freq in tokens.items()
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:405]: filter again

File: lab_3_generate_by_ngrams/main.py [hunk 17]
```diff
@@ -309,6 +451,15 @@ def get_next_token(self, sequence: tuple[int, ...]) -> Optional[list[tuple[int,
...
         In case of corrupt input arguments or methods used return None.
         """
+        if not isinstance(sequence, tuple) or not sequence:
+            return None
+        tokens = self._model.generate_next_token(sequence)
+        if tokens is None:
+            return None
+        if not tokens:
+            return []
+        return sorted([(token, float(freq)) for token, freq in tokens.items()],
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:461]: why do not you cast to float in key?

File: lab_3_generate_by_ngrams/main.py [hunk 18]
```diff
@@ -440,6 +654,35 @@ def load(self, n_gram_size: int) -> Optional[NGramLanguageModel]:
...
+            for token in key:
+                if token.isspace():
+                    encoded.append(0)
+                elif token.isalpha():
+                    ident = self._text_processor.get_id(token.lower())
+                    if not ident:
+                        continue
+                    encoded.append(ident)
+
+            if tuple(encoded) not in n_grams:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:672]: just get

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by artyomtugaryov** [comment]
Very thank you for changes in labels!

File: lab_3_generate_by_ngrams/main.py (hunk 11 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:104]: Changed

File: lab_3_generate_by_ngrams/main.py (hunk 12 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:53]: Changed

File: lab_3_generate_by_ngrams/main.py (hunk 14 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:235]: Rewritten

File: lab_3_generate_by_ngrams/main.py (hunk 18 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:672]: Ok

File: lab_3_generate_by_ngrams/main.py (hunk 13 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Simplified

File: lab_3_generate_by_ngrams/main.py (hunk 15 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Well, I've tried, but it still needed dictionary comprehension🥲

File: lab_3_generate_by_ngrams/main.py (hunk 16 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Done🥲

File: lab_3_generate_by_ngrams/main.py [hunk 19]
```diff
@@ -270,6 +391,27 @@ def run(self, seq_len: int, prompt: str) -> Optional[str]:
...
+            tokens = self._model.generate_next_token(encoded[-n_gram_size + 1:])
+            if not tokens:
+                break
+            max_freq.append(max(tokens.values()))
+            candidates_max = filter(lambda token_freq: token_freq[1] == max_freq[-1],
+                                    tokens.items())
+            encoded += (sorted(candidates_max)[0][0],)
+
+        text = self._text_processor.decode(encoded)
+        if not text:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:412]: again - `text or None`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very very nice!

File: lab_3_generate_by_ngrams/main.py [hunk 20]
```diff
@@ -121,6 +163,15 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(encoded_corpus, tuple):
+            return None
+        decoded_symbols = self._decode(encoded_corpus)
+        if not decoded_symbols:
+            return None
+        decoded_text = self._postprocess_decoded_text(decoded_symbols)
+        if not decoded_text:
```
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:172]: Ok

File: lab_3_generate_by_ngrams/main.py (hunk 19 above)
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py:412]: Ok

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnPruch) [reply lab_3_generate_by_ngrams/main.py]: Changed
